modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -3612,6 +3612,9 @@ def groupby_reduce(
                 self._get_dict_of_block_index(axis ^ 1, numeric_indices).keys()
             )
 
+        if by_parts is not None:
+            # inplace operation
+            self._filter_empties(compute_metadata=False)
```

Comment:
why does `compute_metadata=False`? this will make us to fail in a lot of other cases where metadata is not available:

```python
import pandas
import modin.pandas as pd
from modin.test.storage_formats.pandas.test_internals import (
    construct_modin_df_by_scheme,
)

md_df = construct_modin_df_by_scheme(
    pandas_df=pandas.DataFrame({"a": [1, 1, 2, 2], "b": [3, 4, 5, 6]}),
    partitioning_scheme={"row_lengths": [2, 2], "column_widths": [2]},
)
md_res = md_df.query("a > 1")
grp_obj = md_res.groupby(md_res["a"])
print(grp_obj.count()) # fails with an assertion error
```

<details><summary>traceback</summary>

```
Traceback (most recent call last):
  File "t2.py", line 16, in <module>
    print(grp_obj.count())
  File "modin/modin/logging/logger_decorator.py", line 128, in run_and_log
    return obj(*args, **kwargs)
  File "modin/modin/pandas/groupby.py", line 1165, in count
    return self._wrap_aggregation(
  File "modin/modin/logging/logger_decorator.py", line 128, in run_and_log
    return obj(*args, **kwargs)
  File "modin/modin/pandas/groupby.py", line 1509, in _wrap_aggregation
    query_compiler=qc_method(
  File "modin/modin/logging/logger_decorator.py", line 128, in run_and_log
    return obj(*args, **kwargs)
  File "modin/modin/core/storage_formats/pandas/groupby.py", line 110, in method
    return map_reduce_method(query_compiler, *args, **kwargs)
  File "modin/modin/core/dataframe/algebra/groupby.py", line 86, in <lambda>
    return lambda *args, **kwargs: cls.caller(
  File "modin/modin/core/dataframe/algebra/groupby.py", line 409, in caller
    new_modin_frame = query_compiler._modin_frame.groupby_reduce(
  File "modin/modin/logging/logger_decorator.py", line 128, in run_and_log
    return obj(*args, **kwargs)
  File "modin/modin/core/dataframe/pandas/dataframe/utils.py", line 376, in run_f_on_minimally_updated_metadata
    result = f(self, *args, **kwargs)
  File "modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 3619, in groupby_reduce
    new_partitions = self._partition_mgr_cls.groupby_reduce(
  File "modin/modin/logging/logger_decorator.py", line 128, in run_and_log
    return obj(*args, **kwargs)
  File "modin/modin/core/dataframe/pandas/partitioning/partition_manager.py", line 247, in groupby_reduce
    assert partitions.shape[axis] == by.shape[axis], (
AssertionError: the number of partitions along axis=0 is not equal: 1 != 2
```

</details>

Reply:
we also should keep in mind, that this approach will not guard us from the situations where the 'by' frame itself can have empty partitions:
```python
import pandas
import modin.pandas as pd
from modin.test.storage_formats.pandas.test_internals import (
    construct_modin_df_by_scheme,
)

md_df = construct_modin_df_by_scheme(
    pandas_df=pandas.DataFrame({"a": [1, 1, 2, 2], "b": [3, 4, 5, 6]}),
    partitioning_scheme={"row_lengths": [2, 2], "column_widths": [2]},
)
by_df = construct_modin_df_by_scheme(
    pandas_df=pandas.DataFrame({"a": [1, 1, 2, 2, None, None]}),
    partitioning_scheme={"row_lengths": [2, 2, 2], "column_widths": [2]},
).squeeze()

grp_obj = md_df.groupby(by_df.dropna())
print(grp_obj.count()) # fails with an assertion error
```

Reply:
> why does compute_metadata=False?

@dchigarev I focused on a specific case, that's enough there. If it is necessary to calculate metadata, then most likely it will be necessary to go the other way.

For the workload that I am fixing, this fix is enough. I propose to merge these changes, as they reduce the number of problem cases, and separately continue work on a more complete fix.

Reply:
okay, let's not close the original issue then